service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/IdMapper.java

```diff
@@ -0,0 +1,20 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import lombok.AllArgsConstructor;
+import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
+
+@AllArgsConstructor()
+public class IdMapper {
+    private final RandomIdGeneratorService randomIdGeneratorService;
+    private final Map<String, String> ids = new HashMap<>();
+
+    public String getOrNew(String id) {
```

Comment:
I think we'll need our key to be "{resourceType}/{id}" e.g. "Encounter/6CE8318F-2230-47CE-86CC-5210130A00F5". There is no requirement in FHIR that the id is unique across multiple resource types. If the identifier is a number generated sequentially then collisions will be commonplace. http://hl7.org/fhir/stu3/datatypes.html#id

Maybe the method could be something like `public String createReplacementIdentifier(Resource resource)`?

Reply:
I think the method could get Fhir Reference as parameter because the reference keeps the resource type and the id and while providing mapping we will be mapping more than id of the resource but also the references it has inside. Passing full resource to generate all ids for it, would move much logic that will be already in mapping to find all the references inside and I don't think it should be the responsibility of that class.